### DISCUSSION TITLE
Equivalent of backup CLI from schedule to YAML format

### DISCUSSION THREAD

**Discussion by s4ndalHat** [vmware-tanzu/velero#8345] [Community support Q&A]
Hello, 

Is there any valid value to create backup from schedule like the equivalent of `velero backup create --from-schedule` in a YAML format ? 

I didn't find anything in the API types reference in the documentation.

Thanks!

**Comment by kaovilai** [comment]
`--from-schedule-file` ? something like that?

**Reply by kaovilai** [reply]
not any known flag currently afaict but is this something you wanted?

**Reply by sseago** [reply]
@kaovilai it sounds like they wanted a way to do this avoiding the CLI completely.

**Comment by sseago** [comment]
I'm not sure how this would work. The `velero backup create` CLI is a way to create a backup CR, so the `--from-schedule` tells the CLI how to fill out the backup spec rather than specifying the spec fields via CLI flags.

It sounds like you want to create a backup CR directly from yaml (via `kubectl create -f`), but when you're creating the yaml for a backup, the CLI isn't used at all. The backup CR is created in its entirety and Velero just reconciles it as-is.

In other words, if you want to create the backup CR manually, then just create yaml yourself and pass to `kubectl create/apply` -- if you don't want to specify all of the backup spec fields yourself, that's what the velero CLI is for. If you want to create a backup based on the content of a current backup schedule, the CLI seems like the perfect tool for this.

**Reply by kaovilai** [reply]
The CLI can list available schedule as initial validation, then if does not exists, attempt to parse the string as file name, then failing that, parse as yaml

**Reply by kaovilai** [reply]
in all cases, there will be a valid backup CR creation, as long as we accept that CLI will try to list schedules from cluster first.

**Reply by sseago** [reply]
@kaovilai I think the request was not to "parse schedule as yaml" but "create backup from schedule by creating a yaml file rather than the CLI"

**Reply by kaovilai** [reply]
ok I think I get it now...

PVCs can have dataSource..
backup can also have specSource: scheduleName or something like that.. it is certainly implementable.

**Reply by kaovilai** [reply]
If you don't want to specify all of the backup spec fields yourself, that's what the **<insertSomeNewFieldInBackupSpec> could be for**.

This might have uses for other controllers in cluster or argocd or something else that syncs k8s yaml that don't necesariily have access to velero CLI

**Reply by sseago** [reply]
@kaovilai The biggest problem I see with this is that it's generally considered a bad idea for controllers to modify the spec fields -- but the most straightforward way to implement this would be that upon reconciling a backup with `specSource` included, the controller would copy any spec fields from the source into the current backup that are not already specified in the current backup.

**Reply by s4ndalHat** [reply]
Hello again, what I would like is a configurable value in the YAML format so I can only apply it as a GitOps approach. 
In the CLI, we have the option --from-schedule,

But in YAML, unlike the Restore YAML, I do not have option to specify a "scheduleName" value as example, so when I want to trigger a manual backup, I only need to apply this  YAML file and make my workflow easier

**Reply by kaovilai** [reply]
One wouldn't need to modify specField, as long as it reports what spec was used in status.

**Reply by kaovilai** [reply]
velero.backup.status.scheduleName
velero.backup.status.scheduleSpec

**Reply by sseago** [reply]
@kaovilai that could work, although there are some challenges:
1) We'd have to modify every place in velero code that references backup.spec to check for the presence of bakcup.spec.sourceSpec and possibly pull in that resource and grab from there instead
2) We'd probably need to take 1) and make an approach where "if spec.foo is found, use that, else use sourceSpecObj.template.spec.foo instead -- however, this won't work in cases where the zero value (nil, 0, "", etc.) is a valid value for the field, since it would be ambiguous
3) We run the risk of the external sourceSpec resource being modified between checks. This may not be a big deal, but it is a potential concern.

**Reply by sseago** [reply]
This is potentially a very large change since the backup spec fields are referenced in so many places in the code. If we do decide we want this kind of functionality, I think we really need to implement it in such a way that the backup spec contains all the values we need directly. Since doing this from the backup controller itself seems like a pretty big anti-pattern, perhaps the best alternative would be some sort of new CR that provides some alternate means of creating a Velero backup -- something like a BackupTemplate CR that, when reconciled just results in a backup being created. This would also be a big change, but it's a lot more self-contained than modifying everywhere in Velero that references backup spec fields to do a more complex lookup.

**Reply by s4ndalHat** [reply]
I was just wondering about why this was not implemented, but there is no urgency for me ☺️
Anyway, really glad that you guys are so proactive about it and think about it

Regards,

**Reply by kaovilai** [reply]
Velero can add annotation of backup spec from schedule backupspec during first backup creation to the backup and read from there.. instead of a new status field.

**Reply by kaovilai** [reply]
the backupRequest object can just take spec from the annotation if fromSchedule is found.

**Reply by kaovilai** [reply]
Using annotation would be similar to how they are adding managedFields, or last-applied-configuration annotation.

**Comment by sseago** [comment]
@s4ndalHat This isn't really analogous to the "from-backup" flag on restore -- that's an explicit field in the restore.spec which always references an external object. "from-schedule" is different -- it functions by telling the CLI to copy all of the spec template values from the schedule into the backup spec directly.
